examples/practical-coalton/15_pathnames.lisp

```diff
@@ -0,0 +1,177 @@
+
+  (define-instance (Path String))
+
+  (repr :native cl:pathname)
+  (define-type Pathname)
+
+  (declare pathname->string (Pathname -> String))
+  (define (pathname->string pathname)
+    (lisp String (pathname)
+      (cl:write-to-string pathname)))
+
+  (define-instance (Path Pathname))
+
+  (repr :native cl:pathname)
+  (define-type WildPathname)
+
+  (define-instance (Path WildPathname)))
+
+(coalton-toplevel
+
+  ;; TODO: It's a little unforunate that the Either type is called
+  ;; Result in Coalton, because it makes a usage to represent one of two
+  ;; possible values a little confusing. Is there a more idiomatic way
+  ;; to represent this?
+  (declare list-directory ((Result String Pathname) -> (List Pathname)))
+  (define (list-directory dirname)
+    "Return a list of pathnames for the files in the directory named by DIRNAME."
+    (match dirname
+      ((Err str)
+       (lisp (List Pathname) (str)
+         (pth:list-directory str)))
+      ((Ok pth)
+       (lisp (List Pathname) (pth)
+         (pth:list-directory pth)))))
+
+  (declare file-exists-p ((Path :a) => :a -> (Optional Pathname)))
+  (define (file-exists-p pth)
+    "Return a pathname for the file named by PTH if it exists, otherwise return None."
+    (wrap-lisp-call :a pth:file-exists-p pth))
+
+  (declare directory-pathname-p ((Path :a) => :a -> (Optional :a)))
+  (define (directory-pathname-p pth)
+    (wrap-lisp-call :a pth:directory-pathname-p pth))
+
+  (declare file-pathname-p ((Path :a) => :a -> (Optional :a)))
+  (define (file-pathname-p pth)
+    (wrap-lisp-call :a pth:file-pathname-p pth))
+
+  (declare pathname-as-directory ((Result String Pathname) -> Pathname))
+  (define (pathname-as-directory pth)
+    (match pth
+      ((Err str)
+       (lisp :a (str)
+         (pth:pathname-as-directory str)))
+      ((Ok pth)
+       (lisp :a (pth)
+         (pth:pathname-as-directory pth)))))
+
+  (declare pathname-as-file ((Result String Pathname) -> Pathname))
+  (define (pathname-as-file pth)
+    (match pth
+      ((Err str)
+       (lisp :a (str)
+         (pth:pathname-as-file str)))
+      ((Ok pth)
+       (lisp :a (pth)
+         (pth:pathname-as-file pth)))))
+
+  (declare directory-p ((Path :a) => :a -> (Optional :a)))
+  (define (directory-p pth)
+    "Return a pathname for the directory named by PTH if it is an existing directory.
+Otherwise return None."
+    (wrap-lisp-call :a pth:directory-p pth))
+  
+  (declare file-p ((Path :a) => :a -> (Optional :a)))
+  (define (file-p pth)
+    "Return a pathname for the file named by PTH if it is an existing file.
+Otherwise return None."
+    (wrap-lisp-call :a pth:file-p pth)))
+
+(coalton-toplevel
+  
+  ;; Create an enum type to make the :directories argument to walk-directory
+  ;; more discoverable from the type signature.
+  (define-type WalkDirectoryOptions
+    IncludeDirs
+    ExcludeDirs)
+  
+  ;; TODO: Is there a better way to be able to check if two WalkDirectoryOptions
```

Comment:
Is there a better way to check for equality between elements of an "enum-like" type such as `WalkDirectoryOptions`?